### ISSUE TITLE
Handle xdg-desktop-portal-wlr color-scheme error gracefully

### ISSUE THREAD

**Issue by ghost** [PolyMeilex/sctk-adwaita#33]
Would it be possible to gracefully handle portal on wlroots compositors not finding `color-scheme` with a warning or info message instead of error? From what I read, this isn't planned (https://github.com/emersion/xdg-desktop-portal-wlr/issues/173) to be implemented in xdg-desktop-portal-wlr.

**Comment by PolyMeilex** [comment]
Sorry I don't full understand the request, you mean just lowering the log level for the message? Or is the library behaving in an unexpected way when it can't find the color-scheme value?

**Comment by nohsheef** [comment]
Sorry, I was testing latest wgpu and saw the error. Since it's only a logged (in config.rs) without changing behavior, lowering the level to info or debug would be fine. You may want to turn it into a warning, which I can understand. But at the moment it seems that wlroots compositors would always fail to respond here, and the log message would be visible in daily use, without a way to fix it. Therefore I would vote for info or debug level, if you're okay with that.

**Comment by nohsheef** [comment]
Sorry, I was testing latest wgpu and saw the error. Since it's only logged (in config.rs) without changing behavior, lowering the level to debug is sufficient. You may want to turn it into a warning or info, which I can understand. But at the moment wlroots compositors will always fail to respond here, and the log message will be visible in daily use, without a way to fix it. Therefore I vote for debug level, if you're okay with that.

**Comment by PolyMeilex** [comment]
> without a way to fix it

Well, settings portals are compositor agnostic, you could run one even in raw TTY, so you can simply run one like GTK portal (do not confuse with gnome portal) alongside wlr one, and as far as I remember that's the recommended way to get settings portal as well as system file dialogs support in sway.

But that's unrelated to the issue, will probably lower the log level anyway.

**Comment by kchibisov** [comment]
You've likely haven't exported environment to the portals running. it's a bit racy on sway for example.

**Comment by nohsheef** [comment]
> You've likely haven't exported environment to the portals running. it's a bit racy on sway for example.

I thought I was doing what's necessary by running the official sway package and making sure to source `/etc/sway/config.d/*` as the last line of sway's config. Anything I'm missing to get any sort of portal functionality?

**Comment by nohsheef** [comment]
> You've likely haven't exported environment to the portals running. it's a bit racy on sway for example.

I thought I was doing what's necessary by running the official sway package and making sure to source `/etc/sway/config.d/*` as the last line of sway's config. Anything I'm missing?

**Comment by kchibisov** [comment]
You need this `exec --no-startup-id dbus-update-activation-environment --all` as the first line in the config iirc. it's in the sway wiki.

**Comment by nohsheef** [comment]
Sourcing `/etc/sway/config.d/*` as first line of sway's config works when using portal-gtk. Thanks for the directions. Haven't retested with portal-wlr yet. I would like to disable the document-portal and avoid the unused mount, but that's another issue.

This is what you can find in [/etc/sway/config.d/50-systemd-user.conf](
https://gitlab.archlinux.org/archlinux/packaging/packages/sway/-/raw/main/50-systemd-user.conf):
```
# sway does not set DISPLAY/WAYLAND_DISPLAY in the systemd user environment
# See FS#63021
# Adapted from xorg's 50-systemd-user.sh, which achieves a similar goal.

exec systemctl --user import-environment DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP
exec hash dbus-update-activation-environment 2>/dev/null && \
     dbus-update-activation-environment --systemd DISPLAY WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP
```

What surprised me is that systemd seems to persist the environment even after I log out all sessions for the user, and old variables were still there and conflicting. Looks like more stuff about systemd I have to read. Oh well

**Comment by nohsheef** [comment]
Thanks for the directions, putting this at the start of sway's config works:

```
exec systemctl --user import-environment WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP QT_QPA_PLATFORMTHEME
exec dbus-update-activation-environment --systemd WAYLAND_DISPLAY SWAYSOCK XDG_CURRENT_DESKTOP QT_QPA_PLATFORMTHEME
```

Line 1 is probably unnecessary, but it's what the Arch package has in /etc/sway/config.d/50-systemd-user.conf.

xdg-desktop-portal-wlr didn't work on its own and gave me a resource-unavailable error. Is xdg-desktop-portal-gtk enough or should I complement it with xdg-desktop-portal-wlr?

Either way, the new sway config works for xdg-desktop-portal-gtk to properly answer dbus messages. What surprised me is that systemd seems to persist the environment even after I log out all sessions for the user, and old variables were still there and conflicting.

To solve that and modernize my environment, I migrated to `greetd-tuigreet`. That also works, but now I have to find the right way to start sway with all variables from `~/.config/environment.d` visible. Before, I used to run sway from my login shell. For zsh, it's accomplished via `export $(/usr/lib/systemd/user-environment-generators/30-systemd-environment-d-generator)`. At the moment, apps spawned via fuzzel as sway's launcher menu do not inherit environment variables such as `QT_QPA_PLATFORMTHEME`, despite explicitly listing it. Don't know how to launch sway the right way as the documented greetd recipe from Arch's wiki isn't enough. Maybe sway has to be run as a systemd unit.

This is getting off-topic and I shared the progress as documentation for anyone else coming here after searching the web. I'll stop after this message.
